internal/graph/check_test.go

```diff
@@ -581,3 +582,150 @@ condition condition1(param1: string) {
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if all handlers are falsy, return allowed:false", func(t *testing.T) {
+		resp, err := union(ctx, concurrencyLimit, falseHandler, falseHandler, falseHandler)
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if a handler errors but other handler are truthy, return allowed:true", func(t *testing.T) {
+		depthExceededHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrResolutionDepthExceeded
+		}
+
+		resp, err := union(ctx, concurrencyLimit, depthExceededHandler, trueHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+
+	t.Run("if a handler errors but other handler is falsy, return error and allowed:false", func(t *testing.T) {
+		depthExceededHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrResolutionDepthExceeded
+		}
+
+		resp, err := union(ctx, concurrencyLimit, depthExceededHandler, falseHandler)
+		require.ErrorIs(t, ErrResolutionDepthExceeded, err))
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if a handler errors with cycle but other handler is falsy, return allowed:false with a nil error", func(t *testing.T) {
+		cyclicErrorHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrCycleDetected
+		}
+
+		resp, err := union(ctx, concurrencyLimit, cyclicErrorHandler, falseHandler)
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("should only aggregate DB queries of truthy handlers", func(t *testing.T) {
+		trueHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(1),
+				},
+			}, nil
+		}
+
+		falseHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(9),
+				},
+			}, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, trueHandler, falseHandler) // three handlers
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+		require.Equal(t, uint32(9+1), resp.GetResolutionMetadata().DatastoreQueryCount)
+	})
+
+	t.Run("should aggregate DB queries of all falsy handler", func(t *testing.T) {
+		handler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(3),
+				},
+			}, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, handler, handler, handler) // three handlers
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+		require.Equal(t, uint32(3*3), resp.GetResolutionMetadata().DatastoreQueryCount)
+	})
+
+	t.Run("should return allowed:true if truthy handler evaluated before handler cancels via context", func(t *testing.T) {
```

Comment:
This test is a bit misleading because the handlers can get invoked in different order and they are run concurrently relative to one another. I think it may make more sense to do something like:

```
ctx, cancel := context.WithTimeout(50*time.Millisecond)

resp, err := union(ctx, concurrencyLimit, trueHandler)
require.NoError(t, err)
require.True(t, resp.GetAllowed())
```

This way the context cancellation isn't somehow bound/hidden by the concurrent resolver in the reducer.